lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorStatusManager.java

```diff
@@ -43,8 +43,34 @@ private EvaluatorStatusManager() {
   }
 
   private static boolean isLegal(final EvaluatorState from, final EvaluatorState to) {
-    // TODO
-    return true;
+    if (from == to) {
+      return true;
+    }
+
+    if (!isDoneOrFailedOrKilled(from) && isDoneOrFailedOrKilled(to)) {
+      return true;
+    }
+
+    switch(from) {
+      case ALLOCATED: {
+        return to == EvaluatorState.SUBMITTED;
```

Comment:
Right, the logic is handled here:
!isDoneOrFailedOrKilled(from) && isDoneOrFailedOrKilled(to)